File: hooks/scripts/cmd/preemptive_compaction.go
// PostToolUse hook: warns when context window is near capacity, suggesting /compact.
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/spf13/cobra"
	"solon-hooks/internal/config"
	slctx "solon-hooks/internal/context"
	"solon-hooks/internal/hookio"
)

const (
	compactionWarnThreshold   = 65
	compactionStrongThreshold = 75
	compactionUrgentThreshold = 85
	compactionCooldownMs      = 3 * 60 * 1000 // 3 minutes
	compactionCooldownFile    = "sl-compaction-reminded.json"
)

var preemptiveCompactionCmd = &cobra.Command{
	Use:   "preemptive-compaction",
	Short: "Suggest /compact when context usage exceeds threshold",
	RunE:  runPreemptiveCompaction,
}

func runPreemptiveCompaction(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("preemptive-compaction") {
		os.Exit(0)
	}

	var input hookio.PostToolUseInput
	if err := hookio.ReadInput(&input); err != nil {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	sessionID := input.SessionID
	if sessionID == "" {
		sessionID = os.Getenv("SL_SESSION_ID")
	}
	if sessionID == "" {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	percent, ok := slctx.ReadContextPercent(sessionID)
	if !ok || percent < compactionWarnThreshold {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}

	if !compactionCooldownExpired() {
		hookio.WriteOutput(map[string]interface{}{"continue": true})
		return nil
	}
	writeCompactionCooldown()

	msg := compactionMessage(percent)
	hookio.WriteOutput(map[string]interface{}{
		"continue":          true,
		"additionalContext": msg,
	})
	return nil
}

type compactionCooldownState struct {
	Timestamp int64 `json:"timestamp"`
}

func compactionCooldownExpired() bool {
	path := filepath.Join(os.TempDir(), compactionCooldownFile)
	data, err := os.ReadFile(path)
	if err != nil {
		return true
	}
	var state compactionCooldownState
	if err := json.Unmarshal(data, &state); err != nil {
		return true
	}
	return time.Now().UnixMilli()-state.Timestamp > compactionCooldownMs
}

func writeCompactionCooldown() {
	path := filepath.Join(os.TempDir(), compactionCooldownFile)
	data, _ := json.Marshal(compactionCooldownState{Timestamp: time.Now().UnixMilli()})
	_ = os.WriteFile(path, data, 0644)
}

// compactionMessage returns escalating urgency messages based on context %.
func compactionMessage(percent int) string {
	switch {
	case percent >= compactionUrgentThreshold:
		return fmt.Sprintf(
			"\n[URGENT] Context at %d%% — STOP current work. Run /compact NOW or session will overflow. "+
				"Save any critical context to files before compacting.", percent)
	case percent >= compactionStrongThreshold:
		return fmt.Sprintf(
			"\n[Context Warning] Context at %d%%. Finish current atomic task, then run /compact. "+
				"Avoid spawning new subagents until compacted.", percent)
	default:
		return fmt.Sprintf(
			"\n[Context Notice] Context at %d%%. Plan to /compact soon. Keep responses concise.", percent)
	}
}

File: crates/solon-core/src/hooks/preemptive_compaction.rs
use crate::hooks::{get_str, is_hook_enabled, read_context_percent};
/// PostToolUse hook: warn when context window is near capacity, suggesting /compact.
use anyhow::Result;

const WARN_THRESHOLD: i64 = 65;
const STRONG_THRESHOLD: i64 = 75;
const URGENT_THRESHOLD: i64 = 85;
const COOLDOWN_MS: i64 = 3 * 60 * 1000; // 3 minutes
const COOLDOWN_FILE: &str = "sl-compaction-reminded.json";

pub fn run() -> Result<()> {
    if !is_hook_enabled("preemptive-compaction") {
        std::process::exit(0);
    }

    let input = match crate::hooks::read_hook_input() {
        Ok(v) => v,
        Err(_) => {
            crate::hooks::write_output(&serde_json::json!({"continue": true}));
            return Ok(());
        }
    };

    let session_id = {
        let from_input = get_str(&input, "session_id").to_string();
        if from_input.is_empty() {
            std::env::var("SL_SESSION_ID").unwrap_or_default()
        } else {
            from_input
        }
    };

    if session_id.is_empty() {
        crate::hooks::write_output(&serde_json::json!({"continue": true}));
        return Ok(());
    }

    let percent = match read_context_percent(&session_id) {
        Some(p) if p >= WARN_THRESHOLD => p,
        _ => {
            crate::hooks::write_output(&serde_json::json!({"continue": true}));
            return Ok(());
        }
    };

    if !cooldown_expired() {
        crate::hooks::write_output(&serde_json::json!({"continue": true}));
        return Ok(());
    }
    write_cooldown();

    let msg = compaction_message(percent);
    crate::hooks::write_output(&serde_json::json!({
        "continue": true,
        "additionalContext": msg,
    }));
    Ok(())
}

fn cooldown_path() -> std::path::PathBuf {
    std::env::temp_dir().join(COOLDOWN_FILE)
}

fn cooldown_expired() -> bool {
    let data = match std::fs::read_to_string(cooldown_path()) {
        Ok(d) => d,
        Err(_) => return true,
    };
    let v: serde_json::Value = match serde_json::from_str(&data) {
        Ok(v) => v,
        Err(_) => return true,
    };
    let ts = v.get("timestamp").and_then(|t| t.as_i64()).unwrap_or(0);
    chrono::Utc::now().timestamp_millis() - ts > COOLDOWN_MS
}

fn write_cooldown() {
    let entry = serde_json::json!({ "timestamp": chrono::Utc::now().timestamp_millis() });
    let _ = std::fs::write(
        cooldown_path(),
        serde_json::to_string(&entry).unwrap_or_default(),
    );
}

/// Returns escalating urgency message based on context usage percentage.
fn compaction_message(percent: i64) -> String {
    if percent >= URGENT_THRESHOLD {
        format!(
            "\n[URGENT] Context at {}% — STOP current work. Run /compact NOW or session will overflow. \
             Save any critical context to files before compacting.", percent
        )
    } else if percent >= STRONG_THRESHOLD {
        format!(
            "\n[Context Warning] Context at {}%. Finish current atomic task, then run /compact. \
             Avoid spawning new subagents until compacted.",
            percent
        )
    } else {
        format!(
            "\n[Context Notice] Context at {}%. Plan to /compact soon. Keep responses concise.",
            percent
        )
    }
}
